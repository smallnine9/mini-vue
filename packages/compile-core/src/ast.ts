export const enum NodeTypes {
    ROOT,
    INTERPOLATION,
    SIMPLE_EXPRESSION,
    TEXT,
    ELEMENT
}