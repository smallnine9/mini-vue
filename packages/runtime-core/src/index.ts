export * from './h'
export { createTextVnode } from './vnode'
export { provide, inject } from './apiInject'
export { createRenderer } from './renderer'