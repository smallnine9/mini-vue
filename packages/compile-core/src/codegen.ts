import { NodeTypes } from "./ast"
import { CREATE_ELEMENT_VNODE, helperMapName, TO_DISPLAY_STRING } from "./runtimeHelpers"
import { isString } from '../../shared/index'

export function generate(ast) {
  const context = createCodegenContext()
  const { push } = context
  genFunctionPreamble(ast, context)
  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')
  push(`function ${functionName}(${signature}){`)
  push('return ')
  genNode(ast.codegenNode, context)
  push('}')
  return {
    code: context.code
  }
}

function genFunctionPreamble(node, context) {
  const { push, helper } = context
  const VueBinging = 'Vue'
  const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`
  if (node.helpers.length > 0) {
    push(`const { ${node.helpers.map(aliasHelper).join(',\n')} } = ${VueBinging}`)
  }
  push('\n')
  push('return ')
}

function genNode(node, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break
    case NodeTypes.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context)
      break
  }
}

function genCompoundExpression(node, context) {
  const { push } = context
  const children = node.children
  for(let i = 0; i < children.length; i++) {
    const child = children[i]
    if(isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}

function genElement(node, context) {
  const { push, helper } = context
  const { tag, props, children } = node 
  push(`${helper(CREATE_ELEMENT_VNODE)}(`)
  console.log(children)
  genNodeList(genNullable([tag, props, children]), context)
  push(")")
}

function genNodeList(children, context) {
  const { push } = context
  for(let i = 0; i < children.length; i++) {
    const child = children[i]
    if(isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
    if(i < children.length - 1) {
      push(', ')
    }
  }
}

function genNullable(args) {
  return args.map(arg => arg || 'null')
}

function genExpression(node, context) {
  const { push } = context
  push(node.content)
}

function genInterpolation(node, context) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')
}

function genText(node, context) {
  const { push } = context
  push(`'${node.content}'`)
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }
  return context
}