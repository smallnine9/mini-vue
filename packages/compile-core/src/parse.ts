import { isEnumDeclaration } from "typescript";
import { NodeTypes } from "./ast";

const enum TagType {
    START,
    END
}
export function baseParse(content) {
    const context = createParserContext(content)
    return createRoot(parseChildren(context, []))
}

function createParserContext(content) {
    return {
        source: content
    }
}
function advanceBy(context, length) {
    context.source = context.source.slice(length)
}

function parseChildren(context, ancestors) {
    const nodes: any = []
    while (!isEnd(context, ancestors)) {
        let node
        const s = context.source
        if (s.startsWith('<')) {
            if (/[a-z]/.test(s[1])) {
                node = parseElement(context, ancestors)
            }
        } else if (context.source.startsWith('{{')) {
            node = parseInterpolation(context)
        }
        if (!node) {
            node = parseText(context)
        }
        nodes.push(node)

    }
    return nodes
}

function isEnd(context, ancestors) {
    const s = context.source
    if (s.startsWith('</')) {
        for (let i = ancestors.length - 1; i >= 0; i--) {
            const tag = ancestors[i].tag
            if (tag && startWithEndTagOpen(s, tag)) {
                return true
            }
        }
    }
    return !s
}

function parseInterpolation(context) {
    const openDelimiter = '{{'
    const endDelimiter = '}}'
    const endIndex = context.source.indexOf(
        endDelimiter,
        endDelimiter.length
    )
    advanceBy(context, openDelimiter.length)
    const rawContentLength = endIndex - openDelimiter.length
    const rawContent = parseTextData(context, rawContentLength)
    const content = rawContent.trim()
    advanceBy(context, endDelimiter.length)
    return {
        type: NodeTypes.INTERPOLATION,
        content: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content
        }
    }
}

function parseText(context) {
    const endTag = ['<', '{{']
    let endIndex = context.source.length
    for(let i = 0; i < endTag.length; i++) {
        let index = context.source.indexOf(endTag[i])
        if(index !== -1 && index < endIndex) {
            endIndex = index
        }
    }
    const text = parseTextData(context, endIndex)
    return {
        type: NodeTypes.TEXT,
        content: text
    }
}

function parseTextData(context, length) {
    const text = context.source.slice(0, length)
    advanceBy(context, length)
    return text
}


function createRoot(children) {
    return {
        type: NodeTypes.ROOT,
        children
    }
}

function parseElement(context: any, ancestors: any) {
    //<div></div>
    const element: any = parseTag(context, TagType.START)
    ancestors.push(element)
    element.children = parseChildren(context, ancestors)
    ancestors.pop(element)
    if (startWithEndTagOpen(context.source, element.tag)) {
        parseTag(context, TagType.END)
    } else {
        throw new Error(`缺少结束标签:${element.tag}`)
    }
    return element
}

function startWithEndTagOpen(source, tag) {
    return source.slice(2, 2 + tag.length) === tag
}

function parseTag(context, type: TagType) {
    const match: any = /^<\/?([a-z]*)/i.exec(context.source)
    const tag = match[1]
    advanceBy(context, match[0].length)
    advanceBy(context, 1)
    if (type === TagType.END) return
    return {
        type: NodeTypes.ELEMENT,
        tag
    }
}


