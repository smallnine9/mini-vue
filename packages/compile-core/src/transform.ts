import { NodeTypes } from "./ast"
import { TO_DISPLAY_STRING } from "./runtimeHelpers"

export function transform(root, options = {}) {
    const context = createTransformContext(root, options)
    traverseNode(root, context)
    createRootCodegen(root)
    root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root) {
  const child = root.children[0]
  if(child.type === NodeTypes.ELEMENT) {
    root.codegenNode = root.children[0].codegenNode
  } else {
    root.codegenNode = root.children[0]
  }
}

// 深度优先遍历
function traverseNode(node, context) {
    const { nodeTransforms } = context
    switch(node.type) {
      case NodeTypes.INTERPOLATION:
        context.helper(TO_DISPLAY_STRING)
        break
      case NodeTypes.ROOT:
      case NodeTypes.ELEMENT:
        traverseChildren(node, context)
        break
      default:
        break
    }
    for(let i = 0; i < nodeTransforms.length; i++) {
      const transform = nodeTransforms[i]
      transform(node, context)
    }
}

function traverseChildren(node, context) {
    const { children } = node
    for(let i = 0; i < children.length; i++) {
        traverseNode(children[i], context)
    }
}

function createTransformContext(root, options) {
    const context = {
        root,
        nodeTransforms: options.nodeTransforms || [],
        helpers: new Map(),
        helper(key) {
          context.helpers.set(key, 1)
        }
    }
    return context
}