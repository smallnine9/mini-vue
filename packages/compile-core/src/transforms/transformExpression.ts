import { NodeTypes } from "../ast";
// {
//   type: INTERPOLATION,
//   content: {
//     type: SIMPLE_EXPRESSION,
//     content: message
//   }
// }
export function transformExpression(node) {
  if(node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content)
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`
  return node
}