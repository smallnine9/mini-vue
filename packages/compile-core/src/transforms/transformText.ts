import { isText } from "../utils";
import { NodeTypes } from "../ast";

export function transformText(node) {
  if(node.type === NodeTypes.ELEMENT) {
    const { children } = node
    let currentContainer
    for(let i = 0; i < children.length; i++) {
      let child = children[i]
      if(isText(child)) {
        for(let j = i+1; j < children.length; j++) {
          let next = children[j]
          if(isText(next)) {
            if(!currentContainer) {
              currentContainer = children[i] = {
                type: NodeTypes.COMPOUND_EXPRESSION,
                children: [child]
              }
            }
            currentContainer.children.push(" + ")
            currentContainer.children.push(next)
            children.splice(j, 1)
            j--
          } else {
            currentContainer = undefined;
            break;
          }
        }
      }
    }
  }
}