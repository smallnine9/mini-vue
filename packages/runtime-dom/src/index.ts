import { createRenderer } from '../../runtime-core/src/index'
function patchProp(el: any, key, prevVal, nextVal) {
  const isOn = /^on[A-Z]/.test(key)
  if (isOn) {
    const func = nextVal
    el.addEventListener(key.slice(2).toLocaleLowerCase(), func)
  } else {
    if(nextVal === undefined || nextVal === null) {
      el.removeAttribute(key)
    } else {
      el.setAttribute(key, nextVal)
    }
  }
}

function createElement(type: any) {
  return document.createElement(type)
}

function insert(
  el,
  container,
  anchor
) {
  container.insertBefore(el, anchor)
}

function setElementText(el, text) {
  el.textContent = text
}

function remove(el) {
  const parent = el.parentNode;
  if (parent) {
    parent.removeChild(el);
  }
}

const options = {
  createElement,
  patchProp,
  insert,
  setElementText,
  remove
}

// 暴露给用户的接口
export function createApp(...args) {
  return createRenderer(options).createApp(...args)
}