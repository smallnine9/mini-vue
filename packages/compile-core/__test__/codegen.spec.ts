import { generate } from '../src/codegen'
import { baseParse } from '../src/parse'
import { transform } from '../src/transform'
import { transformExpression } from '../src/transforms/transformExpression'
import { transformElement } from '../src/transforms/transformElement'
import { transformText } from '../src/transforms/transformText'
describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('interpolation', () => {
    const ast = baseParse('{{ message }}')
    transform(ast, {
      nodeTransforms: [transformExpression]
    })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('element', () => {
    const ast = baseParse('<div>hi, {{message}}</div>')
    transform(ast, {
      nodeTransforms: [transformExpression, transformText, transformElement]
    })
    console.log('---------------------------')
    console.log(ast.children[0].children[0].children[2])
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})