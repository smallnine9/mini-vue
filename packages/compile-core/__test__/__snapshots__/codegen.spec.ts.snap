// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen interpolation 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache){return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `
"const {  } = Vue
return function render(_ctx, _cache){return 'hi'}"
`;
